File: 동적계획법/알리바바와 40인의 도둑.py
"""
import sys
# sys.stdin = open("input.txt", 'r')

if __name__ == "__main__":
    n = int(input())
    arr = [list(map(int, input().split())) for _ in range(n)]

    for j in range(1, n):
        arr[0][j] += arr[0][j - 1]

    for i in range(1, n):
        arr[i][0] += arr[i - 1][0]

    for i in range(1, n):
        for j in range(1, n):
            if arr[i - 1][j] < arr[i][j - 1]:
                arr[i][j] += arr[i - 1][j]
            else:
                arr[i][j] += arr[i][j - 1]

    print(arr[n - 1][n - 1])

import sys
# sys.stdin = open("input.txt", 'r')

if __name__ == "__main__":
    n = int(input())
    arr = [list(map(int, input().split())) for _ in range(n)]

    for i in range(n - 2, -1, -1):
        arr[i][n - 1] += arr[i + 1][n - 1]
        arr[n - 1][i] += arr[n - 1][i + 1]

    for i in range(n - 2, -1, -1):
        for j in range(n - 2, -1, -1):
            if arr[i][j + 1] < arr[i + 1][j]:
                arr[i][j] += arr[i][j + 1]
            else:
                arr[i][j] += arr[i + 1][j]

    print(arr[0][0])
"""
import sys
# sys.stdin = open("input.txt", 'r')

def dfs(i, j):
    if memo[i][j] != 0:
        return memo[i][j]
    if i == 0 and j == 0:
        return arr[0][0]
    else:
        if i == 0:
            memo[i][j] = dfs(i, j - 1) + arr[i][j]
        elif j == 0:
            memo[i][j] = dfs(i - 1, j) + arr[i][j]
        else:
            memo[i][j] = min(dfs(i - 1, j), dfs(i, j - 1)) + arr[i][j]

        return memo[i][j]

if __name__ == "__main__":
    n = int(input())
    arr = [list(map(int, input().split())) for _ in range(n)]
    memo = [[0] * n for _ in range(n)]

    print(dfs(n - 1, n - 1))
    

File: 동적계획법/알리바바와 40인의 도둑.cpp
/*
#include <bits/stdc++.h>
using namespace std;

int main(void){
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	// freopen("input.txt", "r", stdin);
	
	int n;
	int i, j;
	cin >> n;
	vector<vector<int> > map(n + 1, vector<int>(n + 1, 0));
	for(i = 1; i <= n; i++){
		for(j = 1; j <= n; j++){
			cin >> map[i][j];
		}
	}
	vector<vector<int> > dis(n + 1, vector<int>(n + 1, 0));
//	for(i = 1; i <= n; i++){
//		dis[i][1] += dis[i - 1][1] + map[i][1];	
//	}
//	for(j = 2; j <= n; j++){
//		dis[1][j] += dis[1][j - 1] + map[1][j];	
//	}
//	
//	for(i = 2; i <= n; i++){
//		for(j = 2; j <= n; j++){
//			dis[i][j] = min(dis[i - 1][j], dis[i][j - 1]) + map[i][j];
//		}
//	}
	for(i = 1; i <= n; i++){
		for(j = 1; j <= n; j++){
			if(i == 1 && j >= 2){
				dis[i][j] += dis[i][j - 1] + map[i][j];	
			}
			else if(i >= 2 && j == 1){
				dis[i][j] += dis[i - 1][j] + map[i][j];
			}
			else{
				dis[i][j] = min(dis[i - 1][j], dis[i][j - 1]) + map[i][j];
			}
		}
	}		
	cout << dis[n][n];
	
	return 0;
}
*/
#include <bits/stdc++.h>
using namespace std;

vector<vector<int> > matrix(21, vector<int>(21));
vector<vector<int> > memo(21, vector<int>(21, 0));

int dfs(int x, int y){
	if(memo[x][y] != 0) return memo[x][y];
	
	if(x == 1 && y == 1) return matrix[1][1];
	else if(x == 1){
		return memo[x][y] = dfs(1, y - 1) + matrix[1][y];
	}
	else if(y == 1){
		return memo[x][y] = dfs(x - 1, 1) + matrix[x][1];
	}
	else{
		return memo[x][y] = min(dfs(x - 1, y), dfs(x, y - 1)) + matrix[x][y];	
	}
}

int main(void){
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	// freopen("input.txt", "r", stdin);
	
	int n;
	int i, j;
	cin >> n;
	
	for(i = 1; i <= n; i++){
		for(j = 1; j <= n; j++){
			cin >> matrix[i][j];
		}
	}	
	
	cout << dfs(n, n);
	
	return 0;
}

